Upload files to a GCS bucket at the end of a refresh

quazi-h opened the pull request:
## Why
We should be uploading the samples TSVs to the user genetic returns workspace for Kathleen on a weekly basis. Right now we are running the Extract, Transform, and Outfile portion, so this last piece is just copying the tsv to another google bucket.
There are a specific changes that will need to be made to cater to the samples workflow that we will need to work through.

[Relevant ticket](https://broadinstitute.atlassian.net/browse/<ticket_id>)

## This PR
Added a new solid that takes DAP survey type and the destination bucket.
It uses the refresh_directory in context along with the provided survey to figure out the files to be uploaded.
Tests are currently not working, need to make a few updates to get them passing but wanted to put up an initial draft for review.

## Checklist
- [ ] Documentation has been updated as needed.

quazi-h committed:
Added a new solid that takes DAP survey type and the destination bucket.
It uses the refresh_directory in context along with the provided survey to figure out the files to be uploaded.
Tests are currently not working, need to make a few updates to get them passing but wanted to put up an initial draft for review.

quazi-h committed:
Adding a return type to write_outfiles - need to pass the list of SurveyTypes to the upload step.
Refactored the outfile so that we are forming the gs:// prefixed string for each outfile.
Commenting out some code to focus on the upload to bucket logic.

aherbst-broad committed:
Switch to using graphs for testability, upgrade dagster (#200)

quazi-h committed:
Merge branch 'master' into qh-dspdc-1911-upload-tsv

quazi-h committed:
Updated the write_outfiles to pass the tsv_directory to the upload_to_gcs solid.
Added a refresh_data_sample graph to pipelines and pulled it into the local_ and dev_ repositories (for testing).
Added the appropriate upload_to_gcs path for prod_repositories (will need updating).
Added a run_local script to set DAGSTER_HOME and run dagit with the local_repositories code.
Updates to get tests passing.

quazi-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
         fan_in_results
     )
+
+    return tuple((fan_in_results, tsv_directory))
```
I had trouble just trying to return "fan_in_results, tsv_directory"
as it kept making my pipeline tests fail with the message:
E           dagster.core.errors.DagsterInvalidConfigError: Error in config for pipeline refresh_data_all
E               Error 1: Missing required config entry "inputs" at path root:solids:upload_to_gcs. Sample config for missing entry: {'inputs': {'tsv_directory': '...'}}
But I wouldn't see the same error for the "fan_in_results" argument. And the scheduled_sample imported from prod_repositories also always passed without issues.

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -232,7 +232,8 @@ def eols_transform_records(config: dict[str, str]) -> dict[str, str]:
 )
-def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: list[DapSurveyType]) -> None:
+def write_outfiles(context: AbstractComputeExecutionContext,
+                   fan_in_results: list[DapSurveyType]) -> tuple: # todo: can I somehow return "list[DapSurveyType], str"?
```
🤔 A dataclass would work here:

```
@dataclass
class FanInResultsWithTsvDir:
     fan_in_results: list[DapSurveyType]
     tsv_dir: str
```
and then just return an instance of that.

quazi-broad replied:
Added the above dataclass to dap_orchestration.types

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+@solid(
+    required_resource_keys={"refresh_directory", "gcs"},
+    config_schema={
+        "upload_dir": String,
```
⛏️ I'd call this `destination_gcs_path` as we're not really uploading to a dir but copying between GCS paths.

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+        "upload_dir": String,
+    }
+)
+def upload_to_gcs(
```
⛏️ Along the same lines, this might be better named `copy_outfiles_to_terra`

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+        inputs: tuple
+    ) -> None:
+    """
+    This solid will take upload the tsvs created from write_outfiles to the provided gs:// bucket. The script
```
⛏️ typo: 
```suggestion
    This solid will copy the tsvs created from write_outfiles to the provided GCS bucket. The script
```

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+        upload_dir = context.solid_config["upload_dir"]
+
+        tsvBucketWithPrefix = parse_gs_path(tsv_dir)
+        uploadBucketWithPrefix = parse_gs_path(upload_dir)
```
⛏️ For consistency, I'd say rename this 
```suggestion
        destBucketWithPrefix = parse_gs_path(upload_dir)
```

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+        destOutfilePrefix = f"{uploadBucketWithPrefix.prefix}/{survey}.tsv"
+
+        source_bucket = storage_client.get_bucket(tsvBucketWithPrefix.bucket)
+        upload_bucket = storage_client.get_bucket(uploadBucketWithPrefix.bucket)
```
Same comment re: naming consistency here:
```suggestion
        dest_bucket = storage_client.get_bucket(destBucketWithPrefix.bucket)
```

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,59 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+        upload_bucket = storage_client.get_bucket(uploadBucketWithPrefix.bucket)
+
+        blob = source_bucket.get_blob(outfilePrefix)
+        if blob.size:
```
⛏️ I'd do the error handling up front:

```suggestion
        if not blob.size:
             raise Failure(...)
      
       <happy path logic>
```

aherbst-broad commented on orchestration/run_local.sh:
```diff
@@ -0,0 +1,11 @@
+  mkdir dagster_home
+fi
+
+DAGSTER_HOME=${WORKING_DIR}/dagster_home dagit -f dap_orchestration/repositories/local_repositories.py
```
⛏️ Newline

aherbst-broad commented on orchestration/dap_orchestration/repositories/dev_repositories.py:
```diff
@@ -1,8 +1,33 @@
+        "io_manager": preconfigure_resource_for_mode(gcs_pickle_io_manager, "dev"),
+        "gcs": google_storage_client,
+    }),
+        refresh_data_sample.to_job(resource_defs={
```
Do we still want this job?

quazi-broad replied:
no, removing this as we were using it to debug.
My last run of that job did successfully run the sample pipeline and copied a file to the destination bucket.

aherbst-broad commented on orchestration/dap_orchestration/pipelines.py:
```diff
@@ -76,4 +14,13 @@ def refresh_data_all() -> None:
+
+
+@graph
+def refresh_data_sample() -> None:
```
This was temporary for testing right? Can it be removed?

aherbst-broad reviewed (changes requested):
Looking good

quazi-h committed:
Apply suggestions from code review

Co-authored-by: aherbst-broad <74209955+aherbst-broad@users.noreply.github.com>

quazi-h committed:
Removing refresh_data_sample job and tests that were used for debugging.
Renamed a few variables / config elements for clarity.
Added a new type FanInResultsWithTsvDir which will be used to pass the survey types and output directory from write_outfiles to copy_outfiles_to_terra.
Refactoring failure logic so errors are handled up front.
Mypy / formatting changes.

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -241,8 +242,57 @@ def write_outfiles(context: AbstractComputeExecutionContext, fan_in_results: lis
+
+        blob = source_bucket.get_blob(outfilePrefix)
+        if not blob.size:
+            raise Failure(f"Error; No {survey} files found in {tsv_dir}")
```
Very small ⛏️ Add a line break for readability
```suggestion
            raise Failure(f"Error; No {survey} files found in {tsv_dir}")
            
```

aherbst-broad committed:
Output terra and non-terra formatted tsvs

aherbst-broad committed:
Fix formatting

aherbst-broad committed:
Fix formatting

aherbst-broad committed:
Fix types

quazi-h committed:
Replaced upload path for samples prod refresh to the official final bucket.
Removed some todos that have been handled.

quazi-h committed:
Update orchestration/dap_orchestration/solids.py

Co-authored-by: aherbst-broad <74209955+aherbst-broad@users.noreply.github.com>

quazi-h committed:
cleanup
